### DISCUSSION TITLE
Jetstream Sequence Number and/or Last Sequence Number in Subject

### DISCUSSION THREAD

**Discussion by FerdinandvHagen** [nats-io/nats-server#4147] [Q&A] [answered]
The Jetstream "RePublish" functionality provides `Nats-Sequence` and `Nats-Last-Sequence` headers among others.

Example:
```
map[Nats-Last-Sequence:[0] Nats-Sequence:[1] Nats-Stream:[TEST] Nats-Subject:[msg.1]]
```

When using a "normal" consumer no such information is provided:
```
sub, err := js.Subscribe("msg.1", func(msg *nats.Msg) {
		fmt.Println(msg.Header) ***--> yields map[]***
})
```

Is there any way to receive the Sequence Number of a message when receiving messages from a Subscriber - or even better the Last Sequence Number?

I'm building an event sourcing setup where I have a large number of ephemeral consumers (devices) that need to catch up and then process messages. Jetstream would act as a event stream that will be limited in length.
From the other discussions I have gathered that using a large number of durable consumers (e.g. bound to a device id) is not advisable. Using Republish and a temporary consumer to catch up old messages seems like a good solution - however without knowing the sequence ids of the messages received from the temporary consumer I'm going to have a hard time sorting messages properly.

**Comment by derekcollison** [comment, marked as answer]
You want to use the messages metadata which has everything you want.

```go
  meta, err := msg.Metadata()
```

The struct from the Go client.
```go
// MsgMetadata is the JetStream metadata associated with received messages.
type MsgMetadata struct {
	Sequence     SequencePair
	NumDelivered uint64
	NumPending   uint64
	Timestamp    time.Time
	Stream       string
	Consumer     string
	Domain       string
}
```

**Reply by FerdinandvHagen** [reply]
Thank you very much! Sorry for that easy question - it was apparently too well hidden for me.

**Reply by alifengineer** [reply]
I want to know what is the difference of count fields, 
```go
count, err := msg.Sub.Delivered()
```
and with metadata.sub.NumDelivered
```go
sub, err := msg.Metadata()
sub.NumDelivered
```

Now i am checking as a number of redelivery messages as a first. I need to get info about how much time this message not processed because if this count more than 3 i want to manual ack this message for this queue.

@derekcollison

**Reply by derekcollison** [reply]
This is more a question for the client team. /cc @piotrpio

**Reply by Jarema** [reply]
Hey!

`msg.Sub.Delivered()` is the count of messages delivered to this subsription (Core NATS subscription that is used internallly by the consumer). I would not use it in this case. This is mostly useful in Core NATS when you use `AutoUnsubscribe` and is not relevant for what you need.

`sub.NumDelivered` is the delivery attempt for this message for given consumer.
This will be incremented if you miss an ack and consumer redelivers the message.
